### ISSUE TITLE
Get all services running on system

### ISSUE THREAD

**Issue by vishal-c-metric** [oshi/oshi#257]
How to get all underlying services running on machine.
For example,
Windows Services.

**Comment by dbwiddis** [comment]
OSHI allows a user to obtain all processes running on a machine (`new SystemInfo().getOperatingSystem().getProcesses()`), but does not detail whether or not those processes are services, as that term isn't necessarily cross-platform or defined similarly on other operating systems (would a daemon be equivalent, or not?).

WMI provides a list of services in the `Win32_Service` class, which could be used to cross-reference a Process by its `ProcessId` field, so you could call
```
Map<String, List<Long>> serviceMap = WmiUtil.selectUint32sFrom(null, "Win32_Service", "ProcessId", null);
List<Long> servicePids = serviceMap.get("ProcessId");
```
Then just check whether `servicePids.contains(Long.valueOf(process.getProcessId()))` to determine whether `process` is a service.

**Comment by vishal-c-metric** [comment]
Thanks dbwiddis :)
That answers the windows services question, but will it work for linux/mac system too or do we need check for services from other util class?

**Comment by dbwiddis** [comment]
It depends on what you call a service.

**Comment by vishal-c-metric** [comment]
Services are process which autostart on system boot.
For example: mongo server
Pardon me if i misunderstoood you :)

**Comment by dbwiddis** [comment]
I did hit enter a bit early, but the point was somewhat made.   While researching this, it's hard to find a good answer.   Not all services start at boot (e.g., I've set up some windows services using Apache Commons Daemon (a.k.a. procrun))  So just finding the startup doesn't work.

For the *nix systems, it gets more complex.  Is it just a daemon (a process with no controlling tty?).   Except in graphical *nix systems, the gui processes run without a tty so you'd have to parse some X output to exclude those.

Is it a process started at boot (check for parent pid = 1, or process uptime close to system uptime)?  A process name ending in 'd'?    Use the `service --status-all` command (it's slow)?   Should you just iterate the contents of /etc/init.d?

Given your definition, the process list does list start time, so you could reasonably find which processes have an up time within some short interval of up time or parent pid.  That may not match the output from other lists.

The bottom line is that the term "service" does not have a consistent cross-platform meaning.

Can you describe your use case for wanting to list services with OSHI?

**Comment by vishal-c-metric** [comment]
I just wanted to get all services whether running or stopped at any particular interval of time on windows/*nix system, if that is possible.
Thanks.

**Comment by dbwiddis** [comment]
I suppose it's possible, just difficult to have things that mean the same thing.   I earlier posted the Windows way to do it. 

The term 'service' may mean something different on the *nix platforms. OS X you can parse `launchctl bstree` but it requires sudo permissions.   Linux has `service --status-all` or `chkconfig --list.  FreeBSD has `service -l` and Solaris has `svcs -a`.  I'm not sure these all have the same meaning as windows services, though, which are much closer to the Linux equivalent of a daemon, which is different than a 'service'.

Different distributions may have different interpretations of what should count on these lists.  Another problem with most of these results is that they simply return a list of process names.  Other than windows they're command-line parsing giving a process name that sometimes takes a while to even process and isn't necessarily easy to directly correlate to running processes.

Philosophically, I don't see much difference in listing stopped services and listing all executable programs/apps... if it's not running, it's just a file on disk (perhaps with an easier syntax using "start" to run the program) and I'm not sure I want to go down that rabbit hole.

The bottom line here, is that just saying "list services" isn't really a useful cross-platform term and should be specifically defined using some metric that's somewhat cross-platform consistent.

**Comment by vishal-c-metric** [comment]
Okay. Thanks for your time and detailed info.
Thanks for creating such awesome library and mainting it. :+1:

**Comment by dbwiddis** [comment]
You can thank @dblock for creating the library and having the awesome foresight of design.  I just helped fill it out.  :)

If there's information that can be added to the existing Process object which helps get you the info you need with Services... or if you have a grand idea of how to do it and want to submit a PR, I'd welcome it!

**Comment by dblock** [comment]
FWIW I like the services abstraction that `brew` adds on OSX, https://github.com/Homebrew/homebrew-services

Linux services are quite standard today. 

Something that abstracts that information could be pretty cool, just my 0.02c.

**Comment by dblock** [comment]
Also, because it's so "opinionated", ie. services aren't as strong basic hardware concepts as CPU, I think it could be a good place for an `oshi-services` separate library.

**Comment by dbwiddis** [comment]
I reopened this issue to keep considering it, or if someone else wants to work on it.  My biggest concern is that, other than Windows, this is essentially just a list of strings.  I don't think MacOS, Linux, or Unix have a strong way of connecting a process name to a process id.    For example, I am currently running two separate applications as "services" on my Mac.  Both simply show up in the process and service list as "java".  I don't know (beyond correlating load, memory, and other factors) which one is which.

**Comment by dbwiddis** [comment]
Also, "stopped" services don't even show up in the process list.  How does this fundamentally differ from an "installed program" or "application"?   If we're going to list services, we should also list applications/programs.

**Comment by henryx** [comment]
In Linux, you can use systemd related utilities (if the distribution uses systemd as init system), but this isn't a simple question, because you can run a service by hand (e.g. KDE Desktop environment launches a per user instance of MySQL/MariaDB after you have logged in). For me, on Posix systems, the only way to get what services are running is matching the parent pid of the process with the pid of the init system (PID 1)

**Comment by Space2Man** [comment]
Hi,

one question from my side. So far we have been running with OSHI 3.6.1 and used the following call to get the list of Windows services (so we are able to monitor and restart them):

Map<String, List<Object>> tmpServices = WmiUtil.selectObjectsFrom(null, "Win32_Service", WMIService.WMIServiceProperties, where, WMIService.WMIServiceValueTypes)

But with 3.8.1 this is not working anymore since you replaced the WmiUtil class. Do you have an example how I could query WMI Win32_Service with 3.8.x onwards?

Thanks and best regards,

    Space

**Comment by dbwiddis** [comment]
I submitted a portion of OSHI's code to the JNA project, and in the process split out much of the previous `WmiUtil` code to `WbemcliUtil`.  3.8.1 has one version of it but there's even one more change upcoming.  This (or something close) should work; the enum is just a plain enum of property name strings:

```
        WmiQuery<WMIService.WMIServiceProperties> serviceQuery = WbemcliUtil.createQuery("Win32_Service WHERE foo=bar",
                WMIService.WMIServiceProperties.class);
        WmiResult<WMIService.WMIServiceProperties> services = WmiUtil.queryWMI(serviceQuery);
        for (int i = 0; i < services.getResultCount(); i++) {
            // process enums here
            doSomethingWith(WmiUtil.getString(services, WMIService.WMIServiceProperties.SERVICENAME, i));
        }
```
In the next version, `createQuery()` will go away and you'll just use `new WmiQuery()` instead.

**Comment by Space2Man** [comment]
Cool, thanks, it's working! Thanks for the quick response!

**Comment by dbwiddis** [comment]
Steps to implement:
- Create a new class for the service, using `OSProcess` as an example. `OSService` seems reasonable but feel free to suggest another one.  As Phil Karlton said, "There are only two hard things in Computer Science: cache invalidation and naming things."
- Decide what attributes to include in the service. I'd suggest, to start, Name, ProcessID, and Status (which could be a String or an enum).  
- Add a `getServices()` method to the `OperatingSystem` interface
- Implement `getServices()` to return an empty array in the `AbstractOperatingSystem` class.  This will prevent all the unimplemented platforms from failing to compile.
- Make a Windows implementation of `getServices()` in the `WindowsOperatingSystem` class.  The code should be something like:

In the class:
```
    enum ServiceProperty {
        NAME, PROCESSID, STATUS;
    }
```
In the `getServices()` method:
```
    WmiQuery<ServiceProperty> serviceQuery = new WmiQuery("Win32_Service", ServiceProperty.class);
    WmiResult<ServiceProperty> services = WmiUtil.queryWMI(serviceQuery);
    for (int i = 0; i < services.getResultCount(); i++) {
        OSService svc = new OSService();
        svc.setName(WmiUtil.getString(services, ServiceProperty.NAME, i));
        // etc...
    } 
```

**Comment by dbwiddis** [comment]
For macOS, run `ExecutingCommand.runNative("launchctl list");` (not `bstree` as stated above and is for much older systems) and parse the output.  The first column is the PID if it's running (a hyphen means it's not), and you can parse the last column (split by dots) for the name (after the seconddot).   The middle column is the last exit code, which is usually 0 and not useful.

Sample output:
```
<danielwiddis> [08-12 20:20] ~/git/oshi $ launchctl list
PID	Status	Label
-	0	com.apple.SafariHistoryServiceAgent
437	0	com.apple.progressd
-	0	com.google.keystone.user.xpcservice
663	0	com.apple.Finder
288	0	com.apple.homed
729	0	com.apple.SafeEjectGPUAgent
-	0	com.apple.quicklook
-	0	com.apple.parentalcontrols.check
-	0	com.apple.PackageKit.InstallStatus
312	0	com.apple.mediaremoteagent
11854	0	com.apple.FontWorker
348	0	com.apple.bird
-	0	com.apple.familycontrols.useragent
-	0	com.apple.AssetCache.agent
358	0	com.apple.universalaccessAuthWarn
279	0	com.apple.nsurlsessiond
-	0	com.apple.syncservices.uihandler
<snip>
-	0	com.apple.mdworker.mail
11869	0	com.apple.mdworker.shared.0E000000-0000-0000-0000-000000000000
-	0	com.apple.mdworker.shared.04000000-0000-0000-0000-000000000000
-	0	com.apple.cvmsCompAgent3425AMD_i386
-	0	com.apple.cvmsCompAgent3425AMD_i386_1
257	0	com.apple.cfprefsd.xpc.agent
-	0	com.apple.DataDetectorsLocalSources
-	0	com.apple.unmountassistant.useragent
-	0	com.apple.java.InstallOnDemand
-	0	com.apple.mbfloagent.0BA0F5BB-88EC-438C-8FD5-9B068F7C345C
<snip>
```
The top of the thread has suggestions for Linux, FreeBSD and Solaris.

**Comment by agithyogendra** [comment]
I have a two questions about these lines of code.

```
WmiQuery<ServiceProperty> serviceQuery = new WmiQuery("Win32_Service", ServiceProperty.class);
WmiResult<ServiceProperty> services = WmiUtil.queryWMI(serviceQuery);
```


1. To make sure, does the first line return the Name, PID, and Status of all the services in the Win32_Service class?

2. What does ` WmiUtil.queryWMI(serviceQuery);` do? I couldn't find much information on this online.

Thanks.

**Comment by dbwiddis** [comment]
Since I wrote those two classes, I'm probably the right one to answer the question!

First, an apology, the code I gave you is outdated.  The newer version uses the `WmiQueryHandler` class.

The javadoc for the relevant classes is [the nested classes here](http://java-native-access.github.io/jna/5.4.0/javadoc/com/sun/jna/platform/win32/COM/WbemcliUtil.html).

The first line just builds a query object that stores the location the data lives.  It stores the class name [`Win32_Service`](https://docs.microsoft.com/en-us/windows/win32/cimwin32prov/win32-service) and also "stores" the names of the columns you want to select, by naming the enum class `ServiceProperty`.  So it translates all this into a WQL Query `SELECT NAME,PROCESSID,STATE FROM Win32_Service`.   But that's it... it's sitting there, ready to be used, as many times as you need it.  

The second line with the queryWMI() method actually runs the query using the COM interface, and stores the result in a WmiResult object. Internally this is an enummap of lists for each column.  You can get the results by giving it the column name ENUM and "row number" which pulls the value from the list.  Each time you run the second line you'll fetch a new set of results, but you can keep the same query (as long as you maintain a reference to it).

Take a look in the `WindowsOperatingSystem` class at the enum `ProcessProperty` and the `WmiQuery` class `PROCESS_QUERY` and the `wmiQueryHandler` field. They're used on lines 414 and 415 and values are fetched from the results on line 423.

**Comment by dbwiddis** [comment]
Some comments on your commit that I see linked to here... the Win32_Service WMI class has a "Status" column that is different than the "State" which I think is what we want.  I would suggest being consistent with the Windows names.   Also I think the "Pathname" may be a field we can collect cross-platform, so you might consider including it also.

**Comment by dbwiddis** [comment]
So while the WMI version is easy to code, it looks like there's a Native API to get the information (and probably what WMI calls under the hood).  This could actually be easier than WMI. :)

JNA already has a [W32ServiceManager](http://java-native-access.github.io/jna/5.4.0/javadoc/com/sun/jna/platform/win32/W32ServiceManager.html) class that you can instantiate, and call [`enumServicesStatusExProcess()`](http://java-native-access.github.io/jna/5.4.0/javadoc/com/sun/jna/platform/win32/W32ServiceManager.html#enumServicesStatusExProcess-int-int-java.lang.String-) to generate an array of [ENUM_SERVICE_STATUS_PROCESS](http://java-native-access.github.io/jna/5.4.0/javadoc/com/sun/jna/platform/win32/Winsvc.ENUM_SERVICE_STATUS_PROCESS.html) structures which have the name, and a nested [SERVICE_STATUS_PROCESS](http://java-native-access.github.io/jna/5.4.0/javadoc/com/sun/jna/platform/win32/Winsvc.SERVICE_STATUS_PROCESS.html) structure with the PID and State....

[Sample code here](https://stackoverflow.com/questions/54882633/get-list-of-services-using-jna), but read the answer to fix the bug.

**Comment by dbwiddis** [comment]
This code might work. Or not. Totally untested, written by hand in this comment based on reading the javadoc:
```
W32ServiceManager sm = new W32ServiceManager();
sm.open(Winsvc.SC_MANAGER_ALL_ACCESS);
Winsvc.ENUM_SERVICE_STATUS_PROCESS[] services =
    enumServicesStatusExProcess(WinNT.SERVICE_WIN32, Winsvc.SERVICE_STATE_ALL, null);
for (Winsvc.ENUM_SERVICE_STATUS_PROCESS svc : services) {
  // Create OSService and do stuff with 
  // svc.lpDisplayName, 
  // svc.ServiceStatusProcess.dwCurrentState, 
  // svc.ServiceStatusProcess.dwProcessId
}
sm.close();
```

**Comment by dbwiddis** [comment]
OK, so a few flaws in the above. No permissions for "All Access" but this does work.  Looks like a PID of 0 applies to not-running services (state 1).  
```
W32ServiceManager sm = new W32ServiceManager();
sm.open(Winsvc.SC_MANAGER_ENUMERATE_SERVICE);
Winsvc.ENUM_SERVICE_STATUS_PROCESS[] services =
    sm.enumServicesStatusExProcess(WinNT.SERVICE_WIN32, Winsvc.SERVICE_STATE_ALL, null);
for (int i = 0; i < services.length; i++) {
  System.out.println(services[i].lpDisplayName+
    "("+services[i].ServiceStatusProcess.dwProcessId+"): "+
    services[i].ServiceStatusProcess.dwCurrentState);
}
sm.close();
```
Output:
```
Adobe Flash Player Update Service(0): 1
Application Experience(1036): 4
Application Layer Gateway Service(0): 1
Application Identity(0): 1
Application Information(0): 1
ASP.NET State Service(0): 1
Windows Audio Endpoint Builder(964): 4
Windows Audio(908): 4
ActiveX Installer (AxInstSV)(0): 1
BitLocker Drive Encryption Service(0): 1
Base Filtering Engine(1404): 4
<snip>
```

**Comment by agithyogendra** [comment]
Thanks for the guidance.

**Comment by agithyogendra** [comment]
The Windows Implementation is done. Working on Linux now.

**Comment by dbwiddis** [comment]
Awesome. I've been digging around for the macOS info for a while, and think I have a suggestion there.  The mac equivalent to services is `launchd` which always runs as PID 1.  The `launchctl` command, however, is not very helpful. There's a user version and a sudo version that return completely different results and would both have to be run.  After further digging for non-sudo results, I found the same code we've already implemented for the Process listing.

Bottom line:   for running macOS services, use the existing "get child processes" code for the children of pid #1.   For stopped services, you can pull information from the `plist` files in the locations [here](https://www.launchd.info/).  It's probably sufficient just to get the name from the filename by doing a directory search, rather than reading each plist file individually.

**Comment by agithyogendra** [comment]
I'm stuck on implementing the getServices method for the Linux OS. I'm looking at the JNA [XAttrUtil ](https://static.javadoc.io/net.java.dev.jna/jna-platform/5.4.0/com/sun/jna/platform/linux/XAttrUtil.html#getXAttr-java.lang.String-java.lang.String-)class currently. Is there another way to do this?

**Comment by agithyogendra** [comment]
I completed the Linux Implementation. I completed it using a ProcessBuilder.

**Comment by dbwiddis** [comment]
We already have the `ExecutingCommand` class which accomplishes the same thing (and `Runtime.exec()` internally calls a ProcessBuilder.

**Comment by agithyogendra** [comment]
I used the ExecutingCommand class in my newest commit. However, I am not sure how to pull the PID from the Class. Should I just leave it as null, since I am able to pull the Service name, and state?

**Comment by dbwiddis** [comment]
I'm not sure what command you're parsing on Linux.  The source code looks like `systemctl list-unit-files` but that doesn't work on my Ubuntu distribution.  `service --status-all` seems to be one way of doing it, but while that does give some notion of status, it doesn't give PIDs.  Obviously if a service isn't running, it has no PID.  

For macOS I'd already decided just to use the "getChildProcesses" command for pid 1, the `launchd` process, and then search for files in the appropriate directories for the "stopped processes".  You can do that for Linux as well, pid 1 is `init`.   I would suggest grabbing the service name strings, then iterating the process list to match running processes, favoring the lowest numbered parent pid to break ties.

**Comment by agithyogendra** [comment]
Linux Implementation is finished. Working on Mac now.

**Comment by agithyogendra** [comment]
Mac Implementation is finished. Working on Unix now.

**Comment by agithyogendra** [comment]
Mac Implementation is finished. Working on Unix now.

**Comment by agithyogendra** [comment]
Unix implementations Are completed
